Save crash reproducer found in experiments.

DonggeLiu opened the pull request:
Fixes #156.
Related: https://github.com/google/oss-fuzz/pull/11700.

Save GS bucket links to the reproducer and other statuses into a JSON file.

DonggeLiu commented:
@oliverchang Please correct me if the sample `benchmark.json` missed anything.
Meanwhile, I will add other code changes to use https://github.com/google/oss-fuzz/pull/11700.

oliverchang commented on report/templates/benchmark.json:
```diff
@@ -0,0 +1,19 @@
+        "benchmark": "{{ benchmark.signature }}",
+        "sample": "{{ sample.id }}",
+        "status": "{{ sample.status }}",
+        "Compiles": "{{ sample.compiles }}",
```
nit: lower  case compiles?

oliverchang commented on report/templates/benchmark.json:
```diff
@@ -0,0 +1,19 @@
+        "coverage": "{{ sample.coverage | percent }}",
+        "coverage_diff": "{{ sample.result.line_coverage_diff }}",
+        "coverage_report": "{{ sample.result.coverage_report_path }}",
+        "stacktrace": "{{ sample.result.stacktrace_link }}",
```
maybe just `output` to generalize this more. This would include both the libFuzzer output + stacktrace (if any).

DonggeLiu replied:
I see. BTW, I am adding a bit more code to https://github.com/google/oss-fuzz/pull/11700 to upload the stacktrace too.
Sorry about the frequent review requests an changes.

oliverchang replied:
Do we  need this?  I assume the  stacktrace/output is  already there somewhere  when we run  `web.py`  ?

DonggeLiu replied:
Would it be better to have a separate item to link to the `libFuzzer` output?
e.g., 
```
"output": "{{ sample.result.output_link }}",
```
This info is already available and need not to be uploaded to the new reproducer bucket directory.

oliverchang reviewed (commented):
benchmark.json looks great, thanks!

DonggeLiu commented:
JOB: https://pantheon.corp.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-172-dg
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-03-19-172-dg-comparison/index.html
BUCKET: https://pantheon.corp.google.com/storage/browser/oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-03-19-172-dg-comparison
BUCKET GS: `gs://oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-03-19-172-dg-comparison`

DonggeLiu commented:
Change since the last exp:
1. Unified the fuzz target binary directory name to `target_binary` for easy access.
2. Automatically load and save benchmark reproducer JOSN.

JOB: https://pantheon.corp.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-172-dg
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-03-20-172-dg-comparison/index.html
BUCKET: https://pantheon.corp.google.com/storage/browser/oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-03-20-172-dg-comparison
BUCKET GS: `gs://oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-03-20-172-dg-comparison`

DonggeLiu commented:
Pretty sure this should work:
PR: https://github.com/google/oss-fuzz-gen/pull/172
JOB: https://pantheon.corp.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-172-dg
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-03-21-172-dg-comparison/index.html
BUCKET: https://pantheon.corp.google.com/storage/browser/oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-03-21-172-dg-comparison
BUCKET GS: `gs://oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-03-21-172-dg-comparison`

oliverchang commented on report/templates/benchmark.json:
```diff
@@ -0,0 +1,22 @@
+        "stacktrace": "{{ sample.stacktrace }}",
+        "target_binary": "{{ sample.target_binary }}",
+        "reproducer": "{{ sample.reproducer }}",
+        "run_log": "{{ sample.run_log }}",
```
What's the difference between stacktrace and run_log? Would it make sense to only have run_log?

DonggeLiu replied:
`run_log` is the fuzzing runtime log from `libFuzzer`, which we already store.
`stackstrace` is the output from `./target_binary reproducer_input`. I added it to simplify future automatic error categorization because of its cleanliness (i.e., no fuzzer output, which can be long but useless for long fuzzing timeout). It may also help to automatically identify if the crash is reproducible.

DonggeLiu replied:
https://github.com/google/oss-fuzz/pull/11700/files#diff-ed7b0455ec9bfd97499697c6be191669e6f378243e24ba14981b4b7f092ffd6cR158

oliverchang replied:
Missed that this was added in https://github.com/google/oss-fuzz/pull/11700.
